src/dbusconnector.cpp

```diff
@@ -0,0 +1,151 @@
+**
+** This file is part of the btrfs-balancer package.
+**
+** You may use this file under the terms of the GNU Lesser General
+** Public License version 2.1 as published by the Free Software Foundation
+** and appearing in the file license.lgpl included in the packaging
+** of this file.
+**
+** This library is free software; you can redistribute it and/or
+** modify it under the terms of the GNU Lesser General Public
+** License version 2.1 as published by the Free Software Foundation
+** and appearing in the file license.lgpl included in the packaging
+** of this file.
+**
+** This library is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+** Lesser General Public License for more details.
+**
+****************************************************************************************/
+
+
+#include "dbusconnector.h"
+#include "maintenance.h"
+
+#include <QCoreApplication>
+#include <QDBusConnection>
+#include <QString>
+#include <QDebug>
+
+namespace
+{
+const QString DBUS_SERVICE("org.nemomobile.BtrfsBalancer");
+const QString DBUS_PATH("/BtrfsBalancer");
+const QString DBUS_INTERFACE("org.nemomobile.BtrfsBalancer");
+
+// timeout in ms after which we service stops after being idle
+const int IDLE_TIMEOUT = 30000;
+}
+
+
+Service::Service(QObject *parent)
+    : QDBusAbstractAdaptor(parent)
+    , m_balancer(new BtrfsBalancer)
+{
+    connect(m_balancer.data(), SIGNAL(status(BtrfsBalancer::Status)),
+            this, SLOT(slotStatusReceived(BtrfsBalancer::Status)));
+    connect(m_balancer.data(), SIGNAL(allocation(qlonglong,qlonglong)),
+            this, SIGNAL(allocation(qlonglong,qlonglong)));
+    connect(m_balancer.data(), SIGNAL(progress(int)),
+            this, SIGNAL(progress(int)));
+    connect(m_balancer.data(), SIGNAL(finished(bool)),
+            this, SIGNAL(finished(bool)));
+
+    connect(&m_idleTimer, SIGNAL(timeout()),
+            this, SLOT(slotIdleTimerTriggered()));
+
+    m_idleTimer.start(IDLE_TIMEOUT);
+}
+
+void Service::checkStatus()
+{
+    m_balancer->checkStatus();
+}
+
+void Service::checkAllocation()
+{
+    m_balancer->checkAllocation();
+}
+
+void Service::balance()
+{
+    m_balancer->balance();
+}
+
+void Service::maintenance(int allocationThreshold,
+                          int batteryThreshold)
+{
+    Maintenance *maintenance = new Maintenance(m_balancer,
+                                               allocationThreshold,
+                                               batteryThreshold);
+    connect(maintenance, SIGNAL(finished()),
+            this, SLOT(slotMaintenanceFinished()));
+    maintenance->start();
+}
+
+void Service::slotStatusReceived(BtrfsBalancer::Status s)
+{
+    emit status(static_cast<int>(s));
+
+    if (s == BtrfsBalancer::BALANCING) {
+        m_idleTimer.stop();
+    } else {
+        m_idleTimer.start(IDLE_TIMEOUT);
+    }
```

Comment:
If the timer was stopped when starting the balancing operation (and restarted only after finish) you would not need to do this.